### PR TITLE
Trello: Promote Init Containers to GA

### PR DESCRIPTION
**Pull request by mburke5678**
For Trello: https://trello.com/c/y5h5MhaB/387-document-node-promote-init-containers-to-ga

Per Derek email: it would be good if we can get a pod spec that uses init-containers demonstrated in this example: https://docs.openshift.org/latest/architecture/core_concepts/pods_and_services.html

there is a sample pod spec that uses an init-containers in the upstream doc here that could be inserted in the referenced example above.
https://kubernetes.io/docs/concepts/workloads/pods/init-containers/

I took the end-user appearing information from
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/container-init.md
and 
https://kubernetes.io/docs/concepts/workloads/pods/init-containers/

Does this section cover the need?
Do I need more information? 
http://file.rdu.redhat.com/mburke/trello-init-containers/trello-init-containers/architecture/core_concepts/pods_and_services.html#init-containers

### CODE REVIEW

**Comment by mburke5678** [comment]
@derekwaynecarr Please assign for tech review as needed.

**Comment by derekwaynecarr** [comment]
LGTM

**Comment by mburke5678** [comment]
@wmengRH PTAL

**Comment by wmengRH** [comment]
1. This part is incorrect "Init containers only support restartPolicy: Always (infinite retries) or restartPolicy: Never (no retries)"
OnFailure supported.
2. in Example 2. Sample Init Container Pod Object Definition (YAML)
There are duplicated "1", the second should be "2".

**Comment by mburke5678** [comment]
@wmengRH 
1. Changed to: "Init containers only support `restartPolicy: Always`, which retries retarting infinitely; `restartPolicy: Never`, which will not retry; and `restartPolicy: OnFailure` which retires with an exponential back-off delay of (10s, 20s, 40s) capped at five minutes." (OnFailure behvior is from the Kubernetes docs).
2. Made change as suggested.

File: architecture/core_concepts/pods_and_services.adoc [hunk 1]
```diff
@@ -160,6 +161,61 @@ link:https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-gui
...
+xref:../../dev_guide/configmaps.adoc#consuming-configmap-in-pods[`restartPolicy`]: 
+
+* Pods configured with `restartPolicy` set to `Never` immediately fail and exit. 
+* Pods configured with `restartPolicy` set to `Always` retry the failing init container with increasing backoff until it succeeds.
+
+[NOTE]
+====
+Init containers only support `restartPolicy: Always`, which retries retarting infinitely; `restartPolicy: Never`, which will not retry; and `restartPolicy: OnFailure` which retires with an exponential back-off delay of (10s, 20s, 40s) capped at five minutes.
+====
+
```
Comment (wmengRH) [inline architecture/core_concepts/pods_and_services.adoc]: All restartPolicy field possible values Always, OnFailure, Never are supported, seems [NOTE] not needed.

File: architecture/core_concepts/pods_and_services.adoc [hunk 2]
```diff
@@ -160,6 +161,61 @@ link:https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-gui
...
+[[init-containers]]
+== Init Containers
+
+An link:https://kubernetes.io/docs/concepts/workloads/pods/init-containers/[init container] is a container in a pod that is started before the pod app containers are started. Init containers can share volumes, perform network operations, and perform computations before the remaining containers start. Init containers can also block or delay the startup of application containers until some precondition is met. 
+
+When a pod starts, after the network and volumes are initialized, the init containers are started in order. Each init container must exit successfully before the next is invoked. If an init container fails to start (due to the runtime) or exits with failure, it is retried according to the pod 
+xref:../../dev_guide/configmaps.adoc#consuming-configmap-in-pods[`restartPolicy`]: 
+
+* Pods configured with `restartPolicy` set to `Never` immediately fail and exit. 
+* Pods configured with `restartPolicy` set to `Always` retry the failing init container with increasing backoff until it succeeds.
```
Comment (wmengRH) [inline architecture/core_concepts/pods_and_services.adoc]: `Always` -> `Always` or `OnFailure`

Comment (mburke5678) [inline architecture/core_concepts/pods_and_services.adoc]: I combined the two lists into a single list:
* `restartPolicy: Always` tries retarting infinitely, with an exponential back-off delay of (10s, 20s, 40s) until the pod is restarted. 
* `restartPolicy: Never` does not try to restart. Pods immediately fail and exit.  
* `restartPolicy: OnFailure` tries restarting with an exponential back-off delay of (10s, 20s, 40s) capped at five minutes.

The upstream Kubernetes doc suggests that OnFailure does not retry until it succeeds. Is that true for OpenShift?

Comment (wmengRH) [inline architecture/core_concepts/pods_and_services.adoc]: for my understanding, 
for init containers, there is no difference between Kubernetes and OpenShift.
and for failed init containers, no difference between Always and OnFailure.
retarting -> restarting

**Comment by wmengRH** [comment]
LGTM

**Comment by mburke5678** [comment]
@openshift/team-documentation PTAL

**Comment by vikram-redhat** [comment]
@bfallonf @gaurav-nelson - PTAL

File: architecture/core_concepts/pods_and_services.adoc [hunk 3]
```diff
@@ -160,6 +161,57 @@ link:https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-gui
 has details about the functionality and purpose of pods.
 ====
 
+[[init-containers]]
```
Comment (gaurav-nelson) [inline architecture/core_concepts/pods_and_services.adoc]: All section titles must have a unique section anchor. `[[init-containers]]` is also used in `architecture/core_concepts/containers_and_images.adoc`

File: architecture/core_concepts/pods_and_services.adoc [hunk 4]
```diff
@@ -160,6 +161,57 @@ link:https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-gui
...
+== Init Containers
+
+An link:https://kubernetes.io/docs/concepts/workloads/pods/init-containers/[init container] is a container in a pod that is started before the pod app containers are started. Init containers can share volumes, perform network operations, and perform computations before the remaining containers start. Init containers can also block or delay the startup of application containers until some precondition is met. 
+
+When a pod starts, after the network and volumes are initialized, the init containers are started in order. Each init container must exit successfully before the next is invoked. If an init container fails to start (due to the runtime) or exits with failure, it is retried according to the pod 
+xref:../../dev_guide/configmaps.adoc#consuming-configmap-in-pods[`restartPolicy`]: 
+
+* `Always` tries restarting infinitely, with an exponential back-off delay of (10s, 20s, 40s) until the pod is restarted.
```
Comment (gaurav-nelson) [inline architecture/core_concepts/pods_and_services.adoc]: Maybe s/infinitely/indefinitely

File: architecture/core_concepts/pods_and_services.adoc [hunk 5]
```diff
@@ -160,6 +161,57 @@ link:https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-gui
...
+When a pod starts, after the network and volumes are initialized, the init containers are started in order. Each init container must exit successfully before the next is invoked. If an init container fails to start (due to the runtime) or exits with failure, it is retried according to the pod 
+xref:../../dev_guide/configmaps.adoc#consuming-configmap-in-pods[`restartPolicy`]: 
+
+* `Always` tries restarting infinitely, with an exponential back-off delay of (10s, 20s, 40s) until the pod is restarted. 
+* `Never` does not try to restart. Pods immediately fail and exit.  
+* `OnFailure` tries restarting with an exponential back-off delay of (10s, 20s, 40s) capped at five minutes.
+
+A pod cannot be ready until all init containers have succeeded. 
+
+The following example outlines a simple pod which has two init containers. The first init container waits for `myservice` and the second waits for `mydb`. Once both containers complete, the Pod starts.
```
Comment (gaurav-nelson) [inline architecture/core_concepts/pods_and_services.adoc]: s/...both containers complete,/...both containers succeed

File: architecture/core_concepts/pods_and_services.adoc [hunk 6]
```diff
@@ -160,6 +161,57 @@ link:https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-gui
...
+    image: busybox
+    command: ['sh', '-c', 'until nslookup myservice; do echo waiting for myservice; sleep 2; done;']
+  - name: init-mydb <2>
+    image: busybox
+    command: ['sh', '-c', 'until nslookup mydb; do echo waiting for mydb; sleep 2; done;']
+----
+
+====
+
+<1> Specifies that the `myservice` container.
```
Comment (gaurav-nelson) [inline architecture/core_concepts/pods_and_services.adoc]: Remove `that`

s/Specifies that the `myservice` container./Specifies the `myservice` container.

File: architecture/core_concepts/pods_and_services.adoc [hunk 7]
```diff
@@ -160,6 +161,57 @@ link:https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-gui
...
+<2> Specifies that the `mydb` container.
```
Comment (gaurav-nelson) [inline architecture/core_concepts/pods_and_services.adoc]: Remove `that`
s/Specifies that the `mydb` container./Specifies the `mydb` container.

File: architecture/core_concepts/pods_and_services.adoc (hunk 4 above)
Comment (gaurav-nelson) [inline architecture/core_concepts/pods_and_services.adoc]: Maybe we can add dashes after `restartPolicy` like:
* `Always` - Tries restarting infinitely, with an exponential back-off delay of (10s, 20s, 40s) until the pod is restarted.
* `Never` - Does not try to restart. Pods immediately fail and exit.
* `OnFailure` - Tries restarting with an exponential back-off delay of (10s, 20s, 40s) capped at five minutes.

**Comment by gaurav-nelson** [comment]
@mburke5678 few comments and suggestions from me, otherwise its all good. :white_check_mark:

**Comment by mburke5678** [comment]
@gaurav-nelson Thanks for the comments! Good catches.

File: architecture/core_concepts/pods_and_services.adoc (hunk 3 above)
Comment (mburke5678) [inline architecture/core_concepts/pods_and_services.adoc]: Nice catch. I wasn't aware of this section in the Containers and Images topic. I changed the link in that section from the Kubernetes doc to this one topic.

**Pull request merged**